### PR TITLE
feat: use docker versioning for docker datasource with javaLTSVersions workaround

### PR DESCRIPTION
**Pull request by t-kulmburg**
## Changes

Split javaLTSVersions workaround into two `packageRules`, one for `docker` and one for `adoptium-java` datasource.
This way updates for dependencies from the `docker` datasource will use the `docker` versioning, and not the `versioning` regex that is currently present. This will now only be used for `adoptium-java`.

## Context

I already described the problem/reason for this MR in https://github.com/renovatebot/renovate/discussions/19516.

**Preventing updates like this:**

**And instead create them like this:**

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
@setchy can you review?

File: lib/config/presets/internal/workarounds.ts [hunk 1]
```diff
@@ -82,7 +82,7 @@ export const presets: Record<string, Preset> = {
         allowedVersions: '/^(?:8|11|17|21|25|29)(?:\\.|$)/',
         description:
           'Limit Java runtime versions to LTS releases. To receive all major releases add `workarounds:javaLTSVersions` to the `ignorePresets` array.',
-        matchDatasources: ['docker', 'adoptium-java'],
+        matchDatasources: ['adoptium-java'],
         matchPackageNames: [
```
Comment (viceice) [inline lib/config/presets/internal/workarounds.ts:86]: our datasource only supports `java`, `java-jre` and `java-jdk` as package names.

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by setchy** [comment]
@t-kulmburg - just wanted to confirm my understanding of the root issue you're looking to resolve.  You would like to maintain the version precision?

Some scenarios would be as follows:

1. `11-jdk` -> `17-jdk`
2. `11.0-jdk` -> `17.0-jdk`
3. `11.0.16-jdk` -> `17.0.5-jdk`
4. `11.0.17_8-jdk` -> `17.0.5_8-jdk`
5. `11.0.17+8` ->  `17.0.5+8`

I had the same question when contributing the initial workaround - https://github.com/renovatebot/renovate/pull/18007#issuecomment-1259721999

Instead of splitting into two packageRules, do you think it is feasible to amend the existing regex?

**Comment by t-kulmburg** [comment]
@setchy - Yes exactly, that is what i am trying to achieve.

I had a look at the regex and tried around a bit, but I am honestly not sure if it is possible fixing this via this regex only.

In the regex `^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$` you already have a `?` for each capturing group, so the regex already matches versions like `17-jdk` or `17.0-jdk` correctly.

My example `eclipse-temurin` has only `17-jdk` and `17.0.5_8-jdk`, so no versions with minor/patch without build and
it seems like it just tries to use a version that fits the full regex as good as possible, and I think this would need different regexes depending on what version precision is currently used:

Using the regexes
`^(?<major>\\d+)?(-(?<compatibility>.*))?$`
`^(?<major>\\d+)?(\\.(?<minor>\\d+))?(-(?<compatibility>.*))?$`
`^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?(-(?<compatibility>.*))?$`
it updates to `17-jdk`

Using the current
`^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$`
it updates to 17.0.5_8-jdk`

**Comment by setchy** [comment]
> @setchy docker versioning doesn't support scenario 4 and 5

yes, correct @viceice.


@t-kulmburg as you pointed out on #19516 - what if the **javaLTSVersions** `allowedVersions` expression was changed from `/^(?:8|11|17|21|25|29)(?:\\.|$)/` to `/^(?:8|11|17|21|25|29)(?:\\.|-|$)/`?.  It seems to do a little better from my testing, with renovate sometimes suggesting only a major update, but other times continuing to do a full semver+ newValue.  Below is an excerpt from the renovate logs from my testing

```
          {
            "depName": "eclipse-temurin",
            "currentValue": "17.0.3_7-jdk-alpine",
            "replaceString": "eclipse-temurin:17.0.3_7-jdk-alpine",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "stage",
            "depIndex": 4,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "17.0.5_8-jdk-alpine",
                "newValue": "17.0.5_8-jdk-alpine",
                "newMajor": 17,
                "newMinor": 0,
                "updateType": "patch",
                "branchName": "renovate/eclipse-temurin-17-x"
              }
            ],
            "warnings": [],
            "versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
            "registryUrl": "https://index.docker.io",
            "currentVersion": "17.0.3_7-jdk-alpine",
            "isSingleVersion": true,
            "fixedVersion": "17.0.3_7-jdk-alpine"
          },
          {
            "depName": "eclipse-temurin",
            "currentValue": "11.0.15_10-jre-alpine",
            "replaceString": "eclipse-temurin:11.0.15_10-jre-alpine",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "stage",
            "depIndex": 5,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "11.0.17_8-jre-alpine",
                "newValue": "11.0.17_8-jre-alpine",
                "newMajor": 11,
                "newMinor": 0,
                "updateType": "patch",
                "branchName": "renovate/eclipse-temurin-11-x"
              },
              {
                "bucket": "major-17",
                "newVersion": "17.0.5_8-jre-alpine",
                "newValue": "17.0.5_8-jre-alpine",
                "newMajor": 17,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/eclipse-temurin-17-x"
              }
            ],
            "warnings": [],
            "versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
            "registryUrl": "https://index.docker.io",
            "currentVersion": "11.0.15_10-jre-alpine",
            "isSingleVersion": true,
            "fixedVersion": "11.0.15_10-jre-alpine"
          },
          {
            "depName": "eclipse-temurin",
            "currentValue": "11",
            "replaceString": "eclipse-temurin:11",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "stage",
            "depIndex": 6,
            "updates": [
              {
                "bucket": "major-17",
                "newVersion": "17",
                "newValue": "17",
                "newMajor": 17,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/eclipse-temurin-17-x"
              }
            ],
            "warnings": [],
            "versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
            "registryUrl": "https://index.docker.io",
            "currentVersion": "11",
            "isSingleVersion": true,
            "fixedVersion": "11"
          },
          {
            "depName": "eclipse-temurin",
            "currentValue": "11-jre",
            "replaceString": "eclipse-temurin:11-jre",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "stage",
            "depIndex": 7,
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "11.0.17_8-jre",
                "newValue": "11.0.17_8-jre",
                "newMajor": 11,
                "newMinor": 0,
                "updateType": "patch",
                "branchName": "renovate/eclipse-temurin-11-x"
              },
              {
                "bucket": "major-17",
                "newVersion": "17.0.5_8-jre",
                "newValue": "17.0.5_8-jre",
                "newMajor": 17,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/eclipse-temurin-17-x"
              }
            ],
            "warnings": [],
            "versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
            "registryUrl": "https://index.docker.io",
            "currentVersion": "11-jre",
            "isSingleVersion": true,
            "fixedVersion": "11-jre"
          },
          {
            "depName": "eclipse-temurin",
            "currentValue": "11-alpine",
            "replaceString": "eclipse-temurin:11-alpine",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "stage",
            "depIndex": 8,
            "updates": [
              {
                "bucket": "major-17",
                "newVersion": "17-alpine",
                "newValue": "17-alpine",
                "newMajor": 17,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/eclipse-temurin-17-x"
              }
            ],
            "warnings": [],
            "versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
            "registryUrl": "https://index.docker.io",
            "currentVersion": "11-alpine",
            "isSingleVersion": true,
            "fixedVersion": "11-alpine"
          },
          {
            "depName": "eclipse-temurin",
            "currentValue": "8-alpine",
            "replaceString": "eclipse-temurin:8-alpine",
            "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
            "datasource": "docker",
            "depType": "stage",
            "depIndex": 9,
            "updates": [
              {
                "bucket": "major-11",
                "newVersion": "11-alpine",
                "newValue": "11-alpine",
                "newMajor": 11,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/eclipse-temurin-11-x"
              },
              {
                "bucket": "major-17",
                "newVersion": "17-alpine",
                "newValue": "17-alpine",
                "newMajor": 17,
                "newMinor": 0,
                "updateType": "major",
                "branchName": "renovate/eclipse-temurin-17-x"
              }
            ],
            "warnings": [],
            "versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
            "registryUrl": "https://index.docker.io",
            "currentVersion": "8-alpine",
            "isSingleVersion": true,
            "fixedVersion": "8-alpine"
          }
```

**Comment by setchy** [comment]
@rarkins @viceice - would it be helpful to add two new versions, one for docker/java (major.minor.patch_build) and another for adoptium (major.minor.patch.build)? Or do we continue with the regex version approach?

**Comment by setchy** [comment]
fyi - raised a separate PR to update the workarounds `allowedVersions` expression - https://github.com/renovatebot/renovate/pull/19605

**Comment by viceice** [comment]
do we still need this PR?

**Comment by setchy** [comment]
> do we still need this PR?

i dont think so

**Comment by t-kulmburg** [comment]
We are currently using using the 2-package-rule variant in our own config, but looks like https://github.com/renovatebot/renovate/pull/19605 is a good improvement already. So no its currently not required any more from my side.

**Pull request closed without merging**